lib/data_stores/mysql_advanced_data_store.py

```diff
@@ -0,0 +1,636 @@
+#!/usr/bin/env python
+# -*- mode: python; encoding: utf-8 -*-
+
+"""An implementation of a data store based on mysql."""
+
+
+import Queue
+import threading
+import thread
+import time
+import MySQLdb
+import re
+
+from MySQLdb import cursors
+
+from grr.lib import aff4
+from grr.lib import config_lib
+from grr.lib import data_store
+from grr.lib import rdfvalue
+from grr.lib import utils
+from grr.lib.data_stores import common
+
+# pylint: disable=nonstandard-exception
+class Error(data_store.Error):
+  """Base class for all exceptions in this module."""
+# pylint: enable=nonstandard-exception
+
+
+class MySQLConnection(object):
+  """A Class to manage MySQL database connections."""
+
+  def __init__(self, queue=None):
+    self.queue = queue
+    try:
+      self._MakeConnection(database=config_lib.CONFIG["Mysql.database_name"])
+    except MySQLdb.OperationalError as e:
+      # Database does not exist
+      if "Unknown database" in str(e):
+        dbh = self._MakeConnection()
+        cursor = dbh.cursor()
+        cursor.execute("Create database `%s`" %
+                       config_lib.CONFIG["Mysql.database_name"])
+
+        self._MakeConnection(database=config_lib.CONFIG["Mysql.database_name"])
+      else:
+        raise
+
+  def _MakeConnection(self, database=""):
+    try:
+      connection_args = dict(
+          user=config_lib.CONFIG["Mysql.database_username"],
+          db=database, charset="utf8",
+          passwd=config_lib.CONFIG["Mysql.database_password"],
+          cursorclass=cursors.DictCursor)
+      if config_lib.CONFIG["Mysql.host"]:
+        connection_args["host"] = config_lib.CONFIG["Mysql.host"]
+      if config_lib.CONFIG["Mysql.port"]:
+        connection_args["port"] = config_lib.CONFIG["Mysql.port"]
+
+      self.dbh = MySQLdb.connect(**connection_args)
+      self.cursor = self.dbh.cursor()
+      self.cursor.connection.autocommit(True)
+
+      return self.dbh
+    except MySQLdb.OperationalError as e:
+      # This is a fatal error, we just raise the top level exception here.
+      if "Access denied" in str(e):
+        raise Error(str(e))
+      raise
+
+  def __enter__(self):
+    return self
+
+  def __exit__(self, unused_type, unused_value, unused_traceback):
+    if self.queue:
+      self.queue.put(self)
+
+  def Execute(self, *args):
+    retries = 50
```

Comment:
Yeah, this is really excessive, but it was there to deal with the constant deadlocking that the delete/insert operations were causing.  Deadlocks are basically down to 0 so this can be safely reduced.